File: simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskView.java
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask;

import android.annotation.TargetApi;
import android.content.Context;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.util.AttributeSet;
import android.widget.EditText;
import android.widget.ScrollView;

import com.zhuinden.simplestack.Bundleable;
import com.zhuinden.simplestackdemoexamplemvp.R;
import com.zhuinden.simplestackdemoexamplemvp.application.Injector;
import com.zhuinden.statebundle.StateBundle;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnTextChanged;

/**
 * Created by Owner on 2017. 01. 26..
 */

public class AddOrEditTaskView
        extends ScrollView
        implements Bundleable {
    public AddOrEditTaskView(Context context) {
        super(context);
        init(context);
    }

    public AddOrEditTaskView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public AddOrEditTaskView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @TargetApi(21)
    public AddOrEditTaskView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    private void init(Context context) {
        if(!isInEditMode()) {
            addOrEditTaskPresenter = Injector.get()
                    .addOrEditTaskPresenter();
        }
    }

    AddOrEditTaskPresenter addOrEditTaskPresenter;

    @OnTextChanged(R.id.add_task_title)
    public void titleChanged(Editable editable) {
        addOrEditTaskPresenter.updateTitle(editable.toString());
    }

    @OnTextChanged(R.id.add_task_description)
    public void descriptionChanged(Editable editable) {
        addOrEditTaskPresenter.updateDescription(editable.toString());
    }

    @BindView(R.id.add_task_title)
    EditText addTaskTitle;

    @BindView(R.id.add_task_description)
    EditText addTaskDescription;

    public void saveTask() {
        addOrEditTaskPresenter.saveTask();
    }

    public void navigateBack() {
        addOrEditTaskPresenter.navigateBack();
    }

    @Override
    public StateBundle toBundle() {
        return addOrEditTaskPresenter.toBundle();
    }

    @Override
    public void fromBundle(@Nullable StateBundle bundle) {
        if(bundle != null) {
            addOrEditTaskPresenter.fromBundle(bundle);
        }
    }

    public void setTitle(String title) {
        addTaskTitle.setText(title);
    }

    public void setDescription(String description) {
        addTaskDescription.setText(description);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        addOrEditTaskPresenter.attachView(this);
    }

    @Override
    protected void onDetachedFromWindow() {
        addOrEditTaskPresenter.detachView(this);
        super.onDetachedFromWindow();
    }
}

File: simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskView.kt
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask

import android.annotation.TargetApi
import android.content.Context
import android.util.AttributeSet
import android.widget.ScrollView
import com.zhuinden.simplestack.navigator.Navigator
import com.zhuinden.simplestackdemoexamplemvp.util.MvpPresenter
import com.zhuinden.simplestackdemoexamplemvp.util.onTextChanged
import kotlinx.android.synthetic.main.path_addoredittask.view.*

/**
 * Created by Owner on 2017. 01. 26..
 */

class AddOrEditTaskView : ScrollView {
    companion object {
        const val CONTROLLER_TAG = "AddOrEditTaskView.Presenter"
    }

    interface Presenter : MvpPresenter<AddOrEditTaskView> {
        fun onTitleChanged(title: String)

        fun onDescriptionChanged(description: String)

        fun onSaveButtonClicked()
    }

    private val addOrEditTaskPresenter by lazy { Navigator.lookupService<Presenter>(context, CONTROLLER_TAG) }

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
    @TargetApi(21)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    fun setTitle(title: String) {
        textAddTaskTitle.setText(title)
    }

    fun setDescription(description: String) {
        textAddTaskDescription.setText(description)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        textAddTaskTitle.onTextChanged { title ->
            addOrEditTaskPresenter.onTitleChanged(title)
        }

        textAddTaskDescription.onTextChanged { description ->
            addOrEditTaskPresenter.onDescriptionChanged(description)
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        addOrEditTaskPresenter.attachView(this)
    }

    override fun onDetachedFromWindow() {
        addOrEditTaskPresenter.detachView(this)
        super.onDetachedFromWindow()
    }

    fun fabClicked() {
        addOrEditTaskPresenter.onSaveButtonClicked()
    }
}
